query/src/main/java/org/infinispan/query/QueryDefinition.java

```diff
@@ -0,0 +1,154 @@
+import org.hibernate.search.spi.IndexedTypeMap;
+import org.hibernate.search.spi.impl.IndexedTypeMaps;
+import org.hibernate.search.spi.impl.PojoIndexedTypeIdentifier;
+import org.infinispan.AdvancedCache;
+import org.infinispan.query.dsl.embedded.impl.EmbeddedQueryEngine;
+import org.infinispan.query.dsl.embedded.impl.HsQueryRequest;
+import org.infinispan.query.dsl.embedded.impl.QueryEngine;
+import org.infinispan.query.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * Stores the query to be executed in a cache in either a String or {@link HSQuery} form together with pagination
+ * and sort information.
+ *
+ * @since 9.2
+ */
+public class QueryDefinition {
+
+   private static final Log log = LogFactory.getLog(QueryDefinition.class, Log.class);
+
+   private String queryString;
+   private HSQuery hsQuery;
+   private int maxResults = 100;
+   private int firstResult;
+   private Set<String> sortableFields;
+   private Class<?> indexedType;
+
+   private final Map<String, Object> namedParameters = new HashMap<>();
+   private transient Sort sort;
+
+   public QueryDefinition(String queryString) {
+      this.queryString = queryString;
+   }
+
+   public QueryDefinition(HSQuery hsQuery) {
+      this.hsQuery = hsQuery;
+   }
+
+   public Optional<String> getQueryString() {
+      return Optional.ofNullable(queryString);
+   }
+
+   private IndexedTypeMap<CustomTypeMetadata> createMetadata() {
+      return IndexedTypeMaps.singletonMapping(PojoIndexedTypeIdentifier.convertFromLegacy(indexedType), () -> sortableFields);
+   }
+
+   protected QueryEngine getQueryEngine(AdvancedCache<?, ?> cache) {
+      return cache.getComponentRegistry().getComponent(EmbeddedQueryEngine.class);
+   }
+
+   public HSQuery getHsQuery(AdvancedCache<?, ?> cache) {
+      if (hsQuery == null) {
+         QueryEngine queryEngine = getQueryEngine(cache);
+         HsQueryRequest hsQueryRequest;
+         if (indexedType != null && sortableFields != null) {
+            IndexedTypeMap<CustomTypeMetadata> metadata = createMetadata();
+            hsQueryRequest = queryEngine.createHsQuery(queryString, metadata, namedParameters);
+         } else {
+            hsQueryRequest = queryEngine.createHsQuery(queryString, null, namedParameters);
+         }
+         this.hsQuery = hsQueryRequest.getHsQuery();
+         this.sort = hsQueryRequest.getSort();
+         hsQuery.firstResult(firstResult);
+         hsQuery.maxResults(maxResults);
+         hsQuery.projection(hsQueryRequest.getProjections());
+      }
+      return hsQuery;
+   }
+
+   public int getMaxResults() {
+      return maxResults;
+   }
+
+   public void setMaxResults(int maxResults) {
+      this.maxResults = maxResults;
+      if (hsQuery != null) {
+         hsQuery.maxResults(maxResults);
+      }
+   }
+
+   public void setNamedParameters(Map<String, Object> params) {
+      if (params != null) params.forEach(this.namedParameters::put);
```

Comment:
Is forEach better that putAll? I suppose we also need to take care of the case when params are null.
```
   if (params == null) {
       namedParameters.clear();
   } else {
       namedParameters.putAll(params); 
   }
```

Reply:
ok